lib/linked_list.rb

```diff
@@ -19,54 +19,134 @@ def initialize
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+      @head = Node.new(value, @head)
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
+      current = @head
+
+      until current.nil?
+        return true if current.data == value
+        current = current.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      max = current.data
+
+      until current.nil?
+        if current.data > max
+          max = current.data
+        end
+        current = current.next
+      end
+
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      min = current.data
+
+      until current.nil?
+        if current.data < min
+          min = current.data
+        end
+        current = current.next
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
     def length
-      raise NotImplementedError
+      count = 0
+      current = @head
+      while !current.nil?
+        count += 1
+        current = current.next
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      count = 0
+      current = @head
+      until count == index
+        return nil if current.nil?
+        current = current.next
+        count += 1
+      end
+      return current.data
     end
 
     # method to print all the values in the linked list
     def visit

```

Comment:
👍   This works, although I would suggest putting all the values on the same line just to make the output more compact.  Something like: `[1, 3, 5, 37, 15]`